src/Repeater/DefaultRepeater.ts

```diff
@@ -0,0 +1,147 @@
+import { logger } from '../Utils';
+import {
+  Repeater,
+  RepeaterEventHandler,
+  RepeaterDeployedEvent,
+  RepeaterEvents
+} from './Repeater';
+import { inject, injectable } from 'tsyringe';
+import io, { Socket } from 'socket.io-client';
+import parser from 'socket.io-msgpack-parser';
+import { hostname } from 'os';
+import { once } from 'events';
+
+export interface DefaultRepeaterOptions {
+  readonly uri: string;
+  readonly token: string;
+}
+
+export const DefaultRepeaterOptions: unique symbol = Symbol(
+  'DefaultRepeaterOptions'
+);
+
+@injectable()
+export class DefaultRepeater implements Repeater {
+  private latestReconnectionError?: Error;
+  private readonly DEFAULT_RECONNECT_TIMES = 3;
+  private socket?: Socket;
+
+  constructor(
+    @inject(DefaultRepeaterOptions)
+    private readonly options: DefaultRepeaterOptions
+  ) {}
+
+  public disconnect() {
+    if (!this.socket) {
+      throw new Error('Not connected');
+    }
+
+    this.socket?.disconnect();
+    this.socket?.removeAllListeners();
+    this.socket = undefined;
+  }
+
+  public async deploy(
+    repeaterId?: string
+  ): Promise<RepeaterDeployedEvent['response']> {
+    if (!this.socket) {
+      throw new Error('Repeater is not connected yet');
+    }
+
+    this.socket.emit('deploy', {
+      repeaterId
+    });
+
+    const [result]: RepeaterDeployedEvent['response'][] = await once(
+      this.socket,
+      'deployed'
+    );
+
+    return result;
+  }
+
+  public connect() {
+    this.socket = io(this.options.uri, {
+      parser,
+      path: '/api/ws/v1',
+      transports: ['websocket'],
+      reconnectionAttempts: this.DEFAULT_RECONNECT_TIMES,
+      auth: {
+        token: this.options.token,
+        domain: hostname()
+      }
+    });
+
+    this.socket.on('connect_error', (error: Error) => {
+      logger.debug(`Unexpected error: %s`, error);
+    });
+
+    logger.debug('Event bus connected to %s', this.options.uri);
+  }
+
+  public on<E extends keyof RepeaterEvents, H extends RepeaterEventHandler<E>>(
+    event: E,
+    handler: H
+  ): void {
+    const eventName: string = event;
+
+    if (event === 'reconnection_failed') {
+      // TODO: Figure out why type is not narrowing
+      this.onReconnectionFailure(handler as any);
+    } else {
+      this.socket.on(eventName, (payload, callback) => {
+        this.processEventHandler(event, payload, handler, callback);
+      });
+    }
+  }
+
+  private onReconnectionFailure<
```

Comment:
Please follow the naming convention, the event handle name must consist of two components, where the first is a noun (e.g. connection) and the second is a verb (e.g .opening or opened). The `on` prefix lacks the clarity